File: WordPress/src/main/java/org/wordpress/android/ui/posts/AddCategoryFragment.java
package org.wordpress.android.ui.posts;

import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatDialogFragment;
import androidx.appcompat.view.ContextThemeWrapper;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;

import org.wordpress.android.R;
import org.wordpress.android.WordPress;
import org.wordpress.android.databinding.AddCategoryBinding;
import org.wordpress.android.fluxc.model.SiteModel;
import org.wordpress.android.fluxc.model.TermModel;
import org.wordpress.android.fluxc.store.TaxonomyStore;
import org.wordpress.android.models.CategoryNode;
import org.wordpress.android.util.ToastUtils;

import java.util.ArrayList;

import javax.inject.Inject;

public class AddCategoryFragment extends AppCompatDialogFragment {
    private SiteModel mSite;
    private AddCategoryBinding mBinding;

    @Inject TaxonomyStore mTaxonomyStore;

    public static AddCategoryFragment newInstance(SiteModel site) {
        AddCategoryFragment fragment = new AddCategoryFragment();
        Bundle bundle = new Bundle();
        bundle.putSerializable(WordPress.SITE, site);
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    @NonNull
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        ((WordPress) requireActivity().getApplication()).component().inject(this);

        initSite(savedInstanceState);

        AlertDialog.Builder builder =
                new MaterialAlertDialogBuilder(new ContextThemeWrapper(getActivity(), R.style.PostSettingsTheme));
        // Get the layout inflater
        LayoutInflater inflater = requireActivity().getLayoutInflater();
        //noinspection InflateParams
        mBinding = AddCategoryBinding.inflate(inflater, null, false);

        loadCategories();

        builder.setView(mBinding.getRoot())
               .setPositiveButton(android.R.string.ok, null)
               .setNegativeButton(android.R.string.cancel, null);

        return builder.create();
    }

    @Override
    public void onStart() {
        super.onStart();
        AlertDialog dialog = (AlertDialog) requireDialog();
        dialog.getButton(DialogInterface.BUTTON_POSITIVE).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (addCategory()) {
                    dismiss();
                }
            }
        });
    }

    private void initSite(Bundle savedInstanceState) {
        if (savedInstanceState == null) {
            if (getArguments() != null) {
                mSite = (SiteModel) getArguments().getSerializable(WordPress.SITE);
            } else {
                mSite = (SiteModel) requireActivity().getIntent().getSerializableExtra(WordPress.SITE);
            }
        } else {
            mSite = (SiteModel) savedInstanceState.getSerializable(WordPress.SITE);
        }

        if (mSite == null) {
            ToastUtils.showToast(requireActivity(), R.string.blog_not_found, ToastUtils.Duration.SHORT);
            getParentFragmentManager().popBackStack();
        }
    }

    private boolean addCategory() {
        String categoryName = mBinding.categoryName.getText().toString();
        CategoryNode selectedCategory = (CategoryNode) mBinding.parentCategory.getSelectedItem();
        long parentId = (selectedCategory != null) ? selectedCategory.getCategoryId() : 0;

        if (categoryName.replaceAll(" ", "").equals("")) {
            mBinding.categoryName.setError(getText(R.string.cat_name_required));
            return false;
        }

        TermModel newCategory = new TermModel(
                TaxonomyStore.DEFAULT_TAXONOMY_CATEGORY,
                categoryName,
                parentId
        );
        ((SelectCategoriesActivity) requireActivity()).categoryAdded(newCategory);

        return true;
    }

    private void loadCategories() {
        CategoryNode rootCategory = CategoryNode.createCategoryTreeFromList(mTaxonomyStore.getCategoriesForSite(mSite));
        ArrayList<CategoryNode> categoryLevels = CategoryNode.getSortedListOfCategoriesFromRoot(rootCategory);
        categoryLevels.add(0, new CategoryNode(0, 0, getString(R.string.top_level_category_name)));
        if (categoryLevels.size() > 0) {
            ParentCategorySpinnerAdapter categoryAdapter =
                    new ParentCategorySpinnerAdapter(getActivity(), R.layout.categories_row_parent, categoryLevels);
            mBinding.parentCategory.setAdapter(categoryAdapter);
        }
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putSerializable(WordPress.SITE, mSite);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mBinding = null;
    }
}

File: WordPress/src/main/java/org/wordpress/android/ui/posts/AddCategoryFragment.kt
package org.wordpress.android.ui.posts

import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatDialogFragment
import androidx.appcompat.view.ContextThemeWrapper
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import org.wordpress.android.R
import org.wordpress.android.WordPress
import org.wordpress.android.databinding.AddCategoryBinding
import org.wordpress.android.fluxc.model.SiteModel
import org.wordpress.android.fluxc.model.TermModel
import org.wordpress.android.fluxc.store.TaxonomyStore
import org.wordpress.android.models.CategoryNode
import org.wordpress.android.util.ToastUtils
import org.wordpress.android.util.extensions.getSerializableCompat
import org.wordpress.android.util.extensions.getSerializableExtraCompat
import javax.inject.Inject

class AddCategoryFragment : AppCompatDialogFragment() {
    private var site: SiteModel? = null
    private var binding: AddCategoryBinding? = null

    @Inject
    lateinit var taxonomyStore: TaxonomyStore

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        (requireActivity().application as WordPress).component().inject(this)
        initSite(savedInstanceState)
        val builder =
            MaterialAlertDialogBuilder(ContextThemeWrapper(activity, R.style.PostSettingsTheme))
        binding = AddCategoryBinding.inflate(layoutInflater, null, false)
        loadCategories()
        builder.setView(binding?.root)
            .setPositiveButton(android.R.string.ok, null)
            .setNegativeButton(android.R.string.cancel, null)

        return builder.create()
    }

    override fun onStart() {
        super.onStart()
        val dialog = requireDialog() as AlertDialog
        dialog.getButton(DialogInterface.BUTTON_POSITIVE).setOnClickListener {
            if (addCategory()) {
                dismiss()
            }
        }
    }

    private fun initSite(savedInstanceState: Bundle?) {
        site = if (savedInstanceState == null) {
            if (arguments != null) {
                requireArguments().getSerializableCompat(WordPress.SITE) as SiteModel?
            } else {
                requireActivity().intent.getSerializableExtraCompat(WordPress.SITE) as SiteModel?
            }
        } else {
            savedInstanceState.getSerializableCompat(WordPress.SITE) as SiteModel?
        }

        if (site == null) {
            ToastUtils.showToast(
                requireActivity(),
                R.string.blog_not_found,
                ToastUtils.Duration.SHORT
            )
            parentFragmentManager.popBackStack()
        }
    }

    private fun addCategory(): Boolean {
        val categoryName = binding?.categoryName?.text.toString()
        val selectedCategory = binding?.parentCategory?.selectedItem as? CategoryNode
        val parentId = selectedCategory?.categoryId

        if (categoryName.replace(" ".toRegex(), "") == "") {
            binding?.categoryName?.error = getText(R.string.cat_name_required)
            return false
        }

        val newCategory = parentId?.let {
            TermModel(
                TaxonomyStore.DEFAULT_TAXONOMY_CATEGORY,
                categoryName,
                it
            )
        }
        (requireActivity() as SelectCategoriesActivity).categoryAdded(newCategory)
        return true
    }

    private fun loadCategories() {
        val rootCategory = site?.let { CategoryNode.createCategoryTreeFromList(taxonomyStore.getCategoriesForSite(it)) }
        val categoryLevels = CategoryNode.getSortedListOfCategoriesFromRoot(rootCategory)
        categoryLevels.add(0, CategoryNode(0, 0, getString(R.string.top_level_category_name)))
        if (categoryLevels.size > 0) {
            val categoryAdapter =
                ParentCategorySpinnerAdapter(
                    activity,
                    R.layout.categories_row_parent,
                    categoryLevels
                )
            binding?.parentCategory?.adapter = categoryAdapter
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putSerializable(WordPress.SITE, site)
    }

    override fun onDestroy() {
        super.onDestroy()
        binding = null
    }

    companion object {
        fun newInstance(site: SiteModel?): AddCategoryFragment {
            val fragment = AddCategoryFragment()
            val bundle = Bundle()
            bundle.putSerializable(WordPress.SITE, site)
            fragment.arguments = bundle
            return fragment
        }
    }
}
